docs/versioning.md

```diff
@@ -0,0 +1,567 @@
+overwritten. This is managed with null versions.
+
+Only one null version should exist at any given time, and it is identified
+in S3 requests and responses with the version id "null".
+
+##### Case 1: Putting Null Versions
+
+With respect to metadata, since the null version is overwritten by subsequent
+null versions, the null version is initially stored in the master key alone,
+as opposed to being stored in the master key and a new version. S3 checks if
+versioning is suspended or has never been configured, and sets the `versionId`
+option to `''` in PUT calls to the metadata engine when creating a new null
+version.
+
+The tables below show the keys resulting from PUT calls to metadata if we put an
+object 'foo' twice, when versioning has not been enabled or is suspended.
+
+(1) PUT foo (first put), versionId: `''`
+
+| key | value |
+|-----|:-----:|
+| foo (null) |   A   |-
+
+(2) PUT foo (second put), versionId: `''`
+
+| key | value |
+|-----|:-----:|
+| foo (null) |   B   |
+
+The S3 API also sets the `isNull` attribute to `true` in the version metadata
+before storing the metadata for these null versions.
+
+##### Case 2: Preserving Existing Null Versions in Versioning-Enabled Bucket
+
+Null versions are preserved when versioning has been enabled or re-enabled, and
+new non-null versions are created.
+
+If the master version is the null version, the S3 API preserves the current null
+version by storing it as a new key `(3A)` in a separate PUT call to metadata,
+prior to overwriting the master version `(3B)`. This implies the null version
+may not necessarily be the latest or master version.
+
+To determine whether the master version is a null version, the S3 API checks
+if the master version's `isNull` property is set to `true`, or if the
+`versionId` attribute of the master version is undefined (indicating it is a
+null version that was put before bucket versioning was configured).
+
+Continuing the example from Case 1, if we enabled versioning and put another
+object, the calls to metadata and resulting keys would resemble the following:
+
+(3A) PUT foo, versionId: `<versionId of master version>` if defined or
+ `<non-versioned object id>`
+
+| key | value |
+|-----|:-----:|
+| foo |   B   |
+| foo.v1 (null) |   B   |
+
+(3B) PUT foo, versioning: `true`
+
+| key | value |
+|-----|:-----:|
+| foo |   C   |
+| foo.v2 |   C   |
+| foo.v1 (null) |   B   |
+
+To prevent issues with concurrent requests, S3 ensures the null version is
+stored with the same version ID by using `versionId` option. S3 sets the
+`versionId` option to the master version's `versionId` metadata attribute value
+during the PUT. This creates a new version with the same version ID of the
+existing null master version.
+
+The null version's `versionId` attribute may be undefined because it was
+generated before the bucket versioning was configured. In that case, a version
+ID is generated using the max timestamp and sequence values possible so that the
```

Comment:
It's the max epoch, part of the `ts` value described for the `versionId` format.